### ISSUE TITLE
building freeorion fails on debian unstable 32bit

### ISSUE THREAD

**Issue by Voker57** [freeorion/freeorion#851]
```
[ 63%] Linking CXX executable ../freeoriond
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `LoadGame(std::string const&, ServerSaveGameData&, std::vector<PlayerSaveGameData, std::allocator<PlayerSaveGameData> >&, Universe&, EmpireManager&, SpeciesManager&, CombatLogManager&, GalaxySetupData&)':
SaveLoad.cpp:(.text+0x3033): undefined reference to `boost::archive::xml_iarchive_impl<boost::archive::xml_iarchive>::load(std::string&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `SaveGame(std::string const&, ServerSaveGameData const&, std::vector<PlayerSaveGameData, std::allocator<PlayerSaveGameData> > const&, Universe const&, EmpireManager const&, SpeciesManager const&, CombatLogManager const&, GalaxySetupData const&, bool)':
SaveLoad.cpp:(.text+0x4ffe): undefined reference to `boost::archive::xml_oarchive_impl<boost::archive::xml_oarchive>::save(std::string const&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::common_oarchive<boost::archive::binary_oarchive>::vsave(boost::archive::class_name_type const&)':
SaveLoad.cpp:(.text._ZN5boost7archive6detail15common_oarchiveINS0_15binary_oarchiveEE5vsaveERKNS0_15class_name_typeE[_ZN5boost7archive6detail15common_oarchiveINS0_15binary_oarchiveEE5vsaveERKNS0_15class_name_typeE]+0x2e): undefined reference to `boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::save(std::string const&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::oserializer<boost::archive::binary_oarchive, PlayerSaveHeaderData>::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11oserializerINS0_15binary_oarchiveE20PlayerSaveHeaderDataE16save_object_dataERNS1_14basic_oarchiveEPKv[_ZNK5boost7archive6detail11oserializerINS0_15binary_oarchiveE20PlayerSaveHeaderDataE16save_object_dataERNS1_14basic_oarchiveEPKv]+0x36): undefined reference to `boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::save(std::string const&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::oserializer<boost::archive::xml_oarchive, PlayerSaveHeaderData>::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11oserializerINS0_12xml_oarchiveE20PlayerSaveHeaderDataE16save_object_dataERNS1_14basic_oarchiveEPKv[_ZNK5boost7archive6detail11oserializerINS0_12xml_oarchiveE20PlayerSaveHeaderDataE16save_object_dataERNS1_14basic_oarchiveEPKv]+0x5d): undefined reference to `boost::archive::xml_oarchive_impl<boost::archive::xml_oarchive>::save(std::string const&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::iserializer<boost::archive::xml_iarchive, PlayerSaveHeaderData>::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11iserializerINS0_12xml_iarchiveE20PlayerSaveHeaderDataE16load_object_dataERNS1_14basic_iarchiveEPvj[_ZNK5boost7archive6detail11iserializerINS0_12xml_iarchiveE20PlayerSaveHeaderDataE16load_object_dataERNS1_14basic_iarchiveEPvj]+0x21): undefined reference to `boost::archive::xml_iarchive_impl<boost::archive::xml_iarchive>::load(std::string&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::oserializer<boost::archive::binary_oarchive, PlayerSaveGameData>::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11oserializerINS0_15binary_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv[_ZNK5boost7archive6detail11oserializerINS0_15binary_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv]+0x4c): undefined reference to `boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::save(std::string const&)'
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11oserializerINS0_15binary_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv[_ZNK5boost7archive6detail11oserializerINS0_15binary_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv]+0xb5): undefined reference to `boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::save(std::string const&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::iserializer<boost::archive::xml_iarchive, PlayerSaveGameData>::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11iserializerINS0_12xml_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj[_ZNK5boost7archive6detail11iserializerINS0_12xml_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj]+0x3b): undefined reference to `boost::archive::xml_iarchive_impl<boost::archive::xml_iarchive>::load(std::string&)'
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11iserializerINS0_12xml_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj[_ZNK5boost7archive6detail11iserializerINS0_12xml_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj]+0xe2): undefined reference to `boost::archive::xml_iarchive_impl<boost::archive::xml_iarchive>::load(std::string&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::oserializer<boost::archive::xml_oarchive, PlayerSaveGameData>::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11oserializerINS0_12xml_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv[_ZNK5boost7archive6detail11oserializerINS0_12xml_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv]+0x72): undefined reference to `boost::archive::xml_oarchive_impl<boost::archive::xml_oarchive>::save(std::string const&)'
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11oserializerINS0_12xml_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv[_ZNK5boost7archive6detail11oserializerINS0_12xml_oarchiveE18PlayerSaveGameDataE16save_object_dataERNS1_14basic_oarchiveEPKv]+0x12a): undefined reference to `boost::archive::xml_oarchive_impl<boost::archive::xml_oarchive>::save(std::string const&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::iserializer<boost::archive::binary_iarchive, PlayerSaveHeaderData>::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11iserializerINS0_15binary_iarchiveE20PlayerSaveHeaderDataE16load_object_dataERNS1_14basic_iarchiveEPvj[_ZNK5boost7archive6detail11iserializerINS0_15binary_iarchiveE20PlayerSaveHeaderDataE16load_object_dataERNS1_14basic_iarchiveEPvj]+0x13): undefined reference to `boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::load(std::string&)'
CMakeFiles/freeoriond.dir/SaveLoad.cpp.o: In function `boost::archive::detail::iserializer<boost::archive::binary_iarchive, PlayerSaveGameData>::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const':
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11iserializerINS0_15binary_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj[_ZNK5boost7archive6detail11iserializerINS0_15binary_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj]+0x24): undefined reference to `boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::load(std::string&)'
SaveLoad.cpp:(.text._ZNK5boost7archive6detail11iserializerINS0_15binary_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj[_ZNK5boost7archive6detail11iserializerINS0_15binary_iarchiveE18PlayerSaveGameDataE16load_object_dataERNS1_14basic_iarchiveEPvj]+0x7e): undefined reference to `boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::load(std::string&)'
../libfreeorioncommon.so: undefined reference to `void boost::log::v2_mt_posix::aux::put_integer<char>(std::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned int, unsigned int, char)'
../libfreeorioncommon.so: undefined reference to `boost::log::v2_mt_posix::aux::code_convert(wchar_t const*, unsigned int, std::string&, std::locale const&)'
../libfreeorioncommon.so: undefined reference to `boost::log::v2_mt_posix::sinks::text_file_backend::consume(boost::log::v2_mt_posix::record_view const&, std::string const&)'
../libfreeorioncommon.so: undefined reference to `boost::filesystem::path_traits::dispatch(boost::filesystem::directory_entry const&, std::string&)'
collect2: error: ld returned 1 exit status
server/CMakeFiles/freeoriond.dir/build.make:459: recipe for target 'freeoriond' failed
make[2]: *** [freeoriond] Error 1
CMakeFiles/Makefile2:340: recipe for target 'server/CMakeFiles/freeoriond.dir/all' failed
make[1]: *** [server/CMakeFiles/freeoriond.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by Vezzra** [comment]
Ok, not really on Linux here, but a few questions: which boost version are you building against? Which C++ standard? C++11? C++98? Dunno, but maybe that can help one of our Linux guys to figure this out...

@apoleon: any ideas? mabe related to #777?

**Comment by apoleon** [comment]
On 09.08.2016 15:01, Vezzra wrote:

> Ok, not really on Linux here, but a few questions: which boost version
> are you building against? Which C++ standard? C++11? C++98? Dunno, but
> maybe that can help one of our Linux guys to figure this out...
> 
> @apoleon https://github.com/apoleon: any ideas? mabe related to #777
> https://github.com/freeorion/freeorion/issues/777?

Sorry, no concrete ideas here but GCC-6 uses -std=gnu++14 as the default
now and Boost is built against the C++11 standard AFAIK.

**Comment by Vezzra** [comment]
Hm... I assume you can build FO on your system?

**Comment by Voker57** [comment]
I can build FO on my desktop which is 64bit. Boost is 1.58, same as desktop. GCC 4.8, I can try with 6.

**Comment by adrianbroher** [comment]
Probably duplicate of #445. Did you try to build the application with static boost libraries?

**Comment by adrianbroher** [comment]
> maybe related to #777?

Certainly not, as the problem occurs during the linker stage.

**Comment by Voker57** [comment]
With GCC 6.1.1 freeoriond builds but freeorion fails with 

```
[ 74%] Linking CXX executable ../../freeorion
../../libGiGi.so: undefined reference to `boost::re_detail::cpp_regex_traits_implementation<char>::transform[abi:cxx11](char const*, char const*) const'
../../libGiGi.so: undefined reference to `boost::re_detail::cpp_regex_traits_implementation<char>::transform_primary[abi:cxx11](char const*, char const*) const'
collect2: error: ld returned 1 exit status
```

**Comment by Voker57** [comment]
Also, 64bit clients fail to connect to 32 bit server.

**Comment by Vezzra** [comment]
Oh... I guess that qualifies as a bug in its own right, and should probably get its own issue. Can you open one?

**Comment by Voker57** [comment]
OK I'll get a log and open one.

**Comment by Vezzra** [comment]
Thx!

**Comment by adrianbroher** [comment]
Stopping the guesswork:
- How did you invoke the compile?
- Can you post a complete compile log?

**Comment by Voker57** [comment]
full build log: http://dump.bitcheese.net/files/wumutus/build.log.gz
compile invocation:

```
cmake ..
make -j2
```

**Comment by LGM-Doyle** [comment]
@Voker57 thanks for the log.

Could you try `make clean`, rebuild everything with `make -j2` and then post the log?

**Comment by Voker57** [comment]
@LGM-Doyle http://dump.bitcheese.net/files/julixyp/build.log.gz

**Comment by Vezzra** [comment]
@Voker57, I guess you forgot to open an issue for what you mentioned here:

> Also, 64bit clients fail to connect to 32 bit server.

Can you do that? Of course I could do it myself, but I guess you can provide a bit more details...

**Comment by Voker57** [comment]
@Vezzra I've been trying to reproduce it, but no luck, must've been a one-off.

**Comment by Vezzra** [comment]
Ok, thanks for reporting back!

**Comment by Vezzra** [comment]
Ok, this is a though one. I need an evaluation by our Linux folks: @adrianbroher, @LGM-Doyle, do you consider this critical for the release? Or can we live with that issue unresolved for 0.4.6?

Will leave the milestone at "Release v0.4.6" for now, waiting for your assessment.

**Comment by adrianbroher** [comment]
> do you consider this critical for the release?

Yes, as this problem will persist for distributions with a more recent boost.

I will set up a local build env with debian to check what we can do here. without having a place to tinker I can only guess.

Also all of those bugs are gcc6.x + boost 1.6x related: this one, #777, #658, #887

**Comment by Vezzra** [comment]
Hm, I see. That's not good, especially if we are dealing with issues that can't be fixed on our side.

Well, I'll leave the "Release v0.4.6" milestone for this issue for now, lets see what you can figure out. As far as the other issues you cited are concerned:
- In #777 [your assessment](https://github.com/freeorion/freeorion/issues/777#issuecomment-238672289) has been: "So far it seems like this can't be worked around in our code". Which, as far as Boost 1.61 is concerned, would also apply to #887 if I understand correctly. Assuming that holds true, it doesn't make sense to block our release for that, does it?
- In #658 there hasn't been any feedback by the ones who originally reported the issue since May 12th. So unless you (or anyone else of our Linux folks) can reproduce their problems and subsequently do anything about them, we're stuck there, and I don't want to hold up the release for that either.

Given that, what shall I do with those other issues? I've removed the "Release v0.4.6" milestone from them for the reasons cited above. Should I tag them again for the release (and thereby block the release) anyway?

I'm aware of the consequences if we don't address those issues, but what are our options?

**Comment by adrianbroher** [comment]
> Assuming that holds true, it doesn't make sense to block our release for that, does it?

This was before @dbenage-cx pointed out the `BOOST_OPTIONAL_CONFIG_USE_OLD_IMPLEMENTATION_OF_OPTIONAL` define. Maybe it's viable for a stopgap solution but no promises here. Also moving around destructors into another scope is a nobrainer. So both of them can maybe fixed.

> So unless you (or anyone else of our Linux folks) can reproduce their problems and subsequently do anything about them, we're stuck there, and I don't want to hold up the release for that either.

I understand, but why the hurry? Forking happens in two weeks and the issues still can be retagged for non release if things don't work out.

**Comment by Vezzra** [comment]
> I understand, but why the hurry? Forking happens in two weeks

Well, the Sept 4th deadline is some kind of latest possible date. _If possible_ I'd like to create the release branch earlier. And up to now it didn't look like we would be able to address those issues, at least not anytime soon.

However, if you think there is some realistic chance we can at least provide some temporary fix for all those issues, I certainly prefer to wait and give it a chance. I _do_ consider all of them quite serious.

So I'm going to set their milestone to "Release v0.4.6" again for now, we can re-evaluate next weekend.

**Comment by adrianbroher** [comment]
> If possible I'd like to create the release branch earlier. And up to now it didn't look like we would be able to address those issues, at least not anytime soon.

Okay, then I misinterpreted your forum post.  I will try to come up with a solution ASAP.

**Comment by Vezzra** [comment]
Oh, sorry if I haven't been clear enough on that. Thanks for trying to get it done sooner... 👍

**Comment by LGM-Doyle** [comment]
I have not followed these compile/link issues because I do not have solutions.

However, since they are currently the main blockers, I will chip in.

Please ignore any unhelpful suggestions.

I have updated the C++11 PR #648 earlier than planned. I was planning on waiting until clang 3.9 was released.  If some of our compile/link problem go away by using C++11, then that suggests where to look for a solution.

**Comment by adrianbroher** [comment]
After using the fix for #777 and making the destructors public freeorion builds properly on a fresh debian install with boost 1.61.0.2 and g++ 6.1.1-12.

Closing as **works for me**
